### PR TITLE
feat(typeahead): support for grouped results (#1274)

### PR DESCRIPTION
**Pull request by gpolychronis**
Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) and [DEVELOPER.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/DEVELOPER.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

### CODE REVIEW

**Comment by gpolychronis** [comment]
@maxokorokov Please review

File: src/typeahead/typeahead-window.ts [hunk 1]
```diff
@@ -116,6 +184,8 @@ export class NgbTypeaheadWindow implements OnInit {
 
   select(item) { this.selectEvent.emit(item); }
 
+  getContext() { return this; }
```
Comment (benouat) [inline src/typeahead/typeahead-window.ts:187]: I am not keen of opening this kind of hole. Not for public API.

File: demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        (c as any).label = category;
+        categories.push(c);
+      }
+      categories[references[category]].push({
+        label: result, idx: idx++
+      });
+    }
+    this.categories = categories;
+  }
+}
```
Comment (benouat) [inline demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts:76]: I think this should be extracted as a standalone file, referenced in the demo. It's not clear for user that the demo contains this component definition. It would appear in the Code tab list

File: demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  'Northern Mariana Islands', 'Ohio', 'Oklahoma', 'Oregon', 'Palau', 'Pennsylvania', 'Puerto Rico', 'Rhode Island',
+  'South Carolina', 'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virgin Islands', 'Virginia',
+  'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'];
+
+@Component({
+  selector: 'ngbd-typeahead-grouping',
+  templateUrl: './typeahead-grouping.html',
+  styles: [`.form-control { width: 300px; }`]
+})
+export class NgbdTypeaheadGrouping {
```
Comment (benouat) [inline demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts:19]: I think our US list of states is not relevant for that kind of demo.
We could inspire ourselves from official Twitter Typeahead.js [multi-datasets example](https://twitter.github.io/typeahead.js/examples#multiple-datasets).

**Review by benouat — COMMENTED** [review]
I left a few comments here.

Overall, I need to think more about this feature. We've done to many times wrong API design choices on the typeahead by the past that we need to put extra effort into that one.

File: src/typeahead/typeahead-window.ts (hunk 1 above)
Comment (maxokorokov) [reply src/typeahead/typeahead-window.ts:187]: It is not the "public" API, but it is required to pass the 'window' to the 'window template' context and then to the 'window item' as an input.
It is a public method, but not a public API (same as `.select()` method).

https://github.com/ng-bootstrap/ng-bootstrap/pull/3711/files#diff-2b05df9eab924379f039993fae7b433eR101

That is the biggest bottleneck in the PR, but I've no idea how to solve it...

File: demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts (hunk 2 above)
Comment (maxokorokov) [reply demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts:76]: I would add that results should not be grouped in the component, but in the search service directly. Component `NgbdTypeaheadGroupingInternal` should just do the display. Also the component should be renamed...

File: demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts (hunk 3 above)
Comment (maxokorokov) [reply demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts:19]: The demo is also broken if you type 2 letters, like "Co" (Colorado, Connecticut). They will be split in 2 groups

**Comment by maxokorokov** [comment]
Also, @gpolychronis, please explain things in future when opening PRs in the description. It helps to understand what is done / not done, what was your thinking, etc. It should be clear what you're trying to do without looking into the code.

Ex. what is that you're introducing, what is the public API, etc.

cc @benouat

File: src/typeahead/typeahead-window.ts (hunk 1 above)
Comment (benouat) [reply src/typeahead/typeahead-window.ts:187]: My point here is to **not** return _this_ !! We return an object containing only what we need to access.

```typescript
getContext() {
return {
  markActive: ...
  select: ... 
  id: ...
  activeIdx: ...
  resultTemplate: ...
}
```

If we need them _live_ we could keep getters so they are executed when accessed, but we don't return _this_. To me, this is not safe/maintainable.

Comment (benouat) [reply src/typeahead/typeahead-window.ts:187]: Other solution, we could also pass every single needed property as one on the `templateContext` itself, instead of forwarding _this_

Then people would write
```html
<.  ..... let-markActive="markActive" let-select="select"  .... >
```

Comment (maxokorokov) [reply src/typeahead/typeahead-window.ts:187]: I don't think there was any intention to give this kind of API for people. What's the point, if we handle all this internally?

Comment (maxokorokov) [reply src/typeahead/typeahead-window.ts:187]: On the other hand I see what you mean. If we have this API, we could use it and then people could use it. And it might be cleaner overall.

Comment (benouat) [reply src/typeahead/typeahead-window.ts:187]: Let me rephrase my concern here:

👍 I am not against the surface API provided by an input `@Input() context` on the component.

👎 I am against providing `this` as a value to it. `this` will move over time, it's opaque, you don't know what it provides. Forcing us to declare what we export as context makes us design better API.

And finally, for pure documentation, we **should** use an interface to declare this context, to avoid such `any` effect as shown below

image

**Comment by gpolychronis** [comment]
> 
> 
> Also, @gpolychronis, please explain things in future when opening PRs in the description. It helps to understand what is done / not done, what was your thinking, etc. It should be clear what you're trying to do without looking into the code.
> 
> Ex. what is that you're introducing, what is the public API, etc.
> 
> cc @benouat

I would like to enable user to write something like that:

<div *ngFor="let category of categories">
        <h3>{{category.label}}</h3>
        <div *ngFor="let item of category">
          <ngb-typeahead-window-item
            [result]="item.label"
            [index]="item.idx"
            [context]="context">
          </ngb-typeahead-window-item>
        </div>
      </div>

I don't like context either. I would like it not be visible to user at all. However as metnioned, the problem remains. How will typeahead instance be injected in items? Any ideas?

File: demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts (hunk 2 above)
Comment (gpolychronis) [reply demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts:76]: There is the issue with change detection. So a new component cannot be avoided. I will try to put in a separate file though.

File: src/typeahead/typeahead-window.ts (hunk 1 above)
Comment (gpolychronis) [reply src/typeahead/typeahead-window.ts:187]: > 
> 
> Let me rephrase my concern here:
> 
> 👍 I am not against the surface API provided by an input `@Input() context` on the component.
> 
> 👎 I am against providing `this` as a value to it. `this` will move over time, it's opaque, you don't know what it provides. Forcing us to declare what we export as context makes us design better API.
> 
> And finally, for pure documentation, we **should** use an interface to declare this context, to avoid such `any` effect as shown below
> 
> image

I would like to make context not visible by users at all. We would like to change what fields belong to context instead of exposing them to users. The only problem is how to inject context in items without users noticing. It isn't meant to be used by users, that's why it has the generic "any". It should not be used by users at all.

**Comment by gpolychronis** [comment]
This is the blocking issue: https://github.com/angular/angular/issues/14935

**Review by benouat — CHANGES_REQUESTED** [review]
@gpolychronis Have a look at TravisCI output, there is a circular dependency in your code. I don't know if the failing UT is due to that, most probably yes.
You'll have to use `forwardRef` I think to overcome this. Search for it on the project to find use case usage of it.

**Comment by gpolychronis** [comment]
@benouat It seems there is an error on Travis side. Am I right, or there is an error on my side? Please chec.

**Comment by gpolychronis** [comment]
UPDATED DESCRIPTION:

In order to enable result grouping, a new template has been added. The new template allows user to customize the layout of the typeahead results.

Technically, there is a problem. The NgbTypeaheadWindow instance must be injected into the items. Right now ngbTypeahead is bound to input element. Input element has no closing tag, so injection via html is not possible.

The solution provided is this: untie ngbTypeahead from input. This way ngbTypeahead can be used with a div element. The layout template can be included inside the this div as long as the input itself. 

At the same time the old way shall be supported too. So ngbTypeahead checks if there is any input child, else it assumes bound element is the input one. Also, if ngbTypeahed is bound to a div, then this div is supposed to include only the input, its label and the layout template.

@maxokorokov @benouat Please review

**Comment by benouat** [comment]
@gpolychronis what is this last force push about ?

**Comment by gpolychronis** [comment]
> 
> 
> @gpolychronis what is this last force push about ?

An attempt to rebuild and check again the tests. It worked.

**Comment by gpolychronis** [comment]
1. Test if it fails gracefully if misused.
2. New PR: update all demo examples.
3. Replace  @ViewChild('input', {static: true}) childInput: ElementRef<HTMLInputElement>; with query
4a. "ngb-typeahead-window-item" -> "ngb-typeahead-item"
4b. "ngb-typeahead-window-item".result -> "ngb-typeahead-item".item

**Comment by gpolychronis** [comment]
> 
> 
>     1. Test if it fails gracefully if misused.
> 
>     2. New PR: update all demo examples.
> 
>     3. Replace  @ViewChild('input', {static: true}) childInput: ElementRef; with query
>        4a. "ngb-typeahead-window-item" -> "ngb-typeahead-item"
>        4b. "ngb-typeahead-window-item".result -> "ngb-typeahead-item".item

Done. @maxokorokov , @benouat please review.

File: demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts (hunk 3 above)
Comment (benouat) [reply demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts:19]: My previous comment here is still valid to me, we should think about another problem domain for this demo. List of states is not good...

**Review by benouat — CHANGES_REQUESTED** [review]
I have seen a lot of progress in the PR 👍
I have reviewed things again and do have some remarks and comments I left here and there.
Please @gpolychronis have a look a them

File: src/carousel/carousel.ts [hunk 4]
```diff
@@ -71,7 +71,8 @@ export class NgbSlide {
           (click)="focus();select(slide.id, NgbSlideEventSource.INDICATOR);"></li>
     </ol>
     <div class="carousel-inner">
-      <div *ngFor="let slide of slides; index as i; count as c" class="carousel-item" [class.active]="slide.id === activeId" [id]="slide.id + '-slide'" role="tabpanel">
+      <div *ngFor="let slide of slides; index as i; count as c" class="carousel-item"
+      [class.active]="slide.id === activeId" [id]="slide.id + '-slide'" role="tabpanel">
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/carousel/carousel.ts:75]: CI cannot work otherwise

**Comment by gpolychronis** [comment]
Just a note: One typeahead template variable "result" is renamed to "item". This will probably break backwards compatibility.

File: src/typeahead/typeahead.ts [hunk 5]
```diff
@@ -45,7 +48,7 @@ const NGB_TYPEAHEAD_VALUE_ACCESSOR = {
  */
 export interface NgbTypeaheadSelectItemEvent<T = any> {
   /**
-   * The item from the result list about to be selected.
+   * The item from the item list about to be selected.
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead.ts [hunk 6]
```diff
@@ -97,7 +99,13 @@ export class NgbTypeahead implements ControlValueAccessor,
    *
    * @since 2.1.0
    */
-  @Input() autocomplete = 'off';
+  @Input()
+  set autocomplete(_autocomplete: string) {
+    this._autocomplete = _autocomplete;
+    if (this._inputRef) {
+      this._inputRef.nativeElement.setAttribute('autocomplete', this._autocomplete);
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead.ts [hunk 7]
```diff
@@ -112,12 +120,12 @@ export class NgbTypeahead implements ControlValueAccessor,
...
 
   /**
-   * If `true`, the first item in the result list will always stay focused while typing.
+   * If `true`, the first item in the item list will always stay focused while typing.
    */
   @Input() focusFirst: boolean;
 
   /**
-   * The function that converts an item from the result list to a `string` to display in the `<input>` field.
+   * The function that converts an item from the item list to a `string` to display in the `<input>` field.
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead.ts [hunk 8]
```diff
@@ -138,7 +146,7 @@ export class NgbTypeahead implements ControlValueAccessor,
   @Input() ngbTypeahead: (text: Observable<string>) => Observable<readonly any[]>;
 
   /**
-   * The function that converts an item from the result list to a `string` to display in the popup.
+   * The function that converts an item from the item list to a `string` to display in the popup.
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead.ts [hunk 9]
```diff
@@ -210,13 +226,20 @@ export class NgbTypeahead implements ControlValueAccessor,
...
   }
 
   ngOnInit(): void {
+    this._inputRef = this.childInput || this._elementRef;
+    this._inputRef.nativeElement.setAttribute('autocomplete', this._autocomplete);
+    this._inputRef.nativeElement.setAttribute('autocapitalize', 'off');
+    this._inputRef.nativeElement.setAttribute('autocorrect', 'off');
+    this._valueChanges = fromEvent<Event>(this._inputRef.nativeElement, 'input')
+                             .pipe(map($event => ($event.target as HTMLInputElement).value));
+
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead.ts [hunk 10]
```diff
@@ -210,13 +226,20 @@ export class NgbTypeahead implements ControlValueAccessor,
...
             this.container === 'body');
       }
     });
   }
 
   ngOnInit(): void {
+    this._inputRef = this.childInput || this._elementRef;
+    this._inputRef.nativeElement.setAttribute('autocomplete', this._autocomplete);
+    this._inputRef.nativeElement.setAttribute('autocapitalize', 'off');
+    this._inputRef.nativeElement.setAttribute('autocorrect', 'off');
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead.ts [hunk 11]
```diff
@@ -405,3 +433,55 @@ export class NgbTypeahead implements ControlValueAccessor,
...
+      [id]="id + '-' + index"
+      [class.active]="index === activeIdx"
+      (mouseenter)="markActive()"
+      (click)="select()">
+        <ng-template [ngTemplateOutlet]="resultTemplate || rt"
+                     [ngTemplateOutletContext]="resultTemplateContext"></ng-template>
+    </button>
+  `
+})
+export class NgbTypeaheadItem {
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead-window.ts [hunk 12]
```diff
@@ -17,25 +16,32 @@ export interface ResultTemplateContext {
   term: string;
 }
 
+/**
+ * The context for the typeahead content template in case you want to override the default one.
+ */
+export interface ContentTemplateContext {
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead-window.ts]: done

File: src/typeahead/typeahead-window.ts [hunk 13]
```diff
@@ -17,25 +16,32 @@ export interface ResultTemplateContext {
...
-      <button type="button" class="dropdown-item" role="option"
-        [id]="id + '-' + idx"
-        [class.active]="idx === activeIdx"
-        (mouseenter)="markActive(idx)"
-        (click)="select(result)">
-          <ng-template [ngTemplateOutlet]="resultTemplate || rt"
-          [ngTemplateOutletContext]="{result: result, term: term, formatter: formatter}"></ng-template>
-      </button>
+    <ng-template #defaultContentTemplate let-r="results">
+      <ng-template ngFor [ngForOf]="r" let-item let-idx="index">
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead-window.ts]: done

File: src/typeahead/typeahead-window.ts [hunk 14]
```diff
@@ -17,25 +16,32 @@ export interface ResultTemplateContext {
...
+        <ngb-typeahead-item
+          [item]="item"
+          [index]="idx">
+        </ngb-typeahead-item>
+      </ng-template>
     </ng-template>
+    <ng-template [ngTemplateOutlet]="contentTemplate || defaultContentTemplate"
+                 [ngTemplateOutletContext]="{results: results}"></ng-template>
```
[the remark this answers is not in the document]
Comment (gpolychronis) [reply src/typeahead/typeahead-window.ts]: done

**Comment by finau** [comment]
Any update on the progress team?  we need this feature please.

**Comment by Nosfistis** [comment]
What happened to this PR? Any chance to progress it or any other typeahead group approach?
